design_docs/2023_08_02_shader_parameter_format.md

```diff
@@ -0,0 +1,100 @@
+# Shader parameter format
+
+What kind of format should we use for transferring shader parameters in json? I've been thinking about two designs.
+
+Let's look at a single type that we would like to represent. This is what it looks like in wgsl:
+
+```wgsl
+struct MyStruct {
+    value_a: vec4<f32>,
+    value_b: u32
+}
+```
+
+This is what it would look like in rust:
+
+```rust
+#[repr(C)]
+struct MyStruct {
+    value_a: [f32; 4],
+    value_b: u32,
+}
+```
+
+## First idea
+
+We only transport an array of primitives, as they are laid out in the type after stripping out all structs, vecs, matrices and arrays:
+
+```json
+[
+    {
+        "type": "f32",
+        "val": 0.0,
+    },
+
+    {
+        "type": "f32",
+        "val": 0.0,
+    },
+
+    {
+        "type": "f32",
+        "val": 0.0,
+    },
+
+    {
+        "type": "f32",
+        "val": 0.0,
+    },
+
+    {
+        "type": "u32",
+        "val": 42,
+    },
+]
+```
+
+## Second idea
+
+We include information about the structure of the data:
+
+```json
+{
+    "type": "struct",
+    "val": [
+        {
+            "type": "list",
```

Comment:
```suggestion
            "type": "list",
            "name": "field_name"
```
We don't need field names, but it would be useful for validation and error messages. Relaying on order of the fields only would be hard for users